src/libraries/System.Private.CoreLib/src/System/Reflection/FunctionPointerInfo.cs

```diff
@@ -0,0 +1,94 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Reflection
+{
+    internal sealed partial class FunctionPointerInfo
+    {
+        private const string CallingConventionTypePrefix = "System.Runtime.CompilerServices.CallConv";
+        private readonly RuntimeFunctionPointerParameterInfo _returnInfo;
+        private readonly RuntimeFunctionPointerParameterInfo[] _parameterInfos;
+        private Type[]? _callingConventions;
+
+        public FunctionPointerParameterInfo ReturnParameter => _returnInfo;
+        public FunctionPointerParameterInfo[] ParameterInfos => CloneArray(_parameterInfos);
+
+        public Type[] GetCallingConventions()
+        {
+            if (_callingConventions == null)
+            {
+                Type[] customModifiers = _returnInfo.GetOptionalCustomModifiers();
+                List<Type>? list = null;
+                bool foundCallingConvention = false;
+
+                for (int i = 0; i < customModifiers.Length; i++)
+                {
+                    Type type = customModifiers[i];
+                    if (type.FullName!.StartsWith(CallingConventionTypePrefix))
+                    {
+                        list ??= new List<Type>();
+                        list.Add(type);
+
+                        if (type == typeof(CallConvCdecl) ||
+                            type == typeof(CallConvFastcall) ||
+                            type == typeof(CallConvStdcall) ||
+                            type == typeof(CallConvThiscall))
+                        {
+                            foundCallingConvention = true;
+                        }
+                    }
+                }
+
+                // Normalize the calling conventions.
+                if (!foundCallingConvention)
+                {
+                    Type? callConv = null;
+
+                    switch (CallingConvention)
+                    {
+                        case MdSigCallingConvention.C:
+                            callConv = typeof(CallConvCdecl);
+                            break;
+                        case MdSigCallingConvention.FastCall:
+                            callConv = typeof(CallConvFastcall);
+                            break;
+                        case MdSigCallingConvention.StdCall:
+                            callConv = typeof(CallConvStdcall);
+                            break;
+                        case MdSigCallingConvention.ThisCall:
+                            callConv = typeof(CallConvThiscall);
+                            break;
+                    }
+
+                    if (callConv != null)
+                    {
+                        list ??= new List<Type>();
+                        list.Add(callConv);
+                        _returnInfo.GetOptionalCustomModifiersList().Add(callConv);
```

Comment:
There's something unpleasant about this call to `Add`.  It's modifying a `private` member of `RoFunctionPointerParameterInfo/`RuntimeFunctionPointerParameterInfo` and there's basically no indication in the implementation of those classes that that can happen.

If the field has to be mutated, it might be better to have an internal function `AddCallConvModOpt` (or some better name) on the parameter info class so that the mutation is local to the class.

Also doesn't this mean that the return value of `functionPointerSig.ReturnParameter.GetOptionalCustomModifiers()` changes based on whether you called `functionPointerSig.GetCallingConventions()` ?  Is that really how that public API should behave?

Reply:
Also, does Mono have an equivalent of an internal "Signature" class like Core? Currently this PR uses the Signature class since the native Type holds onto it, and because it's a convenient way to get access to the necessary information. However, based on a couple factors I'm thinking of removing that copy of the signature and just adding extern methods to get the information directly without a signature (e.g. GetAllTypesFromFunctionPointer, GetRawCallingConventionFromFunctionPointer, GetCustomModifiersFromFunctionPointer).

Reply:
> The public return value should be the same whether or not which one comes first -- I'll take a look.

the public return value is:
```csharp
public override Type[] GetOptionalCustomModifiers()
        {
            return _optionalModifiers == null ?
                _signature.GetCustomModifiers(_position + 1, required: false) :
                _optionalModifiers.ToArray();
        }
```

which means if `_optionalModifiers` gets updated, the next time `GetOptionalCustomModifiers` is called it will return a different value.

> Basically, this code is attempting defer "normalizing" the calling convention information so those modifiers appear as Types in GetOptionalCustomModifiersList() even though they may not have been encoded that way (they may have just been in the CallKind; no modifier).

yep, the desired behavior makes sense, but modifying a `private` field from outside the class in which it's defined is a bug waiting to happen.

Reply:
> Also, does Mono have an equivalent of an internal "Signature" class like Core? Currently this PR uses the Signature class since the native Type holds onto it, and because it's a convenient way to get access to the necessary information. However, based on a couple factors I'm thinking of removing that copy of the signature and just adding extern methods to get the information directly without a signature (e.g. GetAllTypesFromFunctionPointer, GetRawCallingConventionFromFunctionPointer, GetCustomModifiersFromFunctionPointer).

Mono doesn't have a signature surfaced to managed right now.  My plan was roughly as you suggest: just make FunctionPointerInfo a wrapper around a type handle and add internal calls to retrieve the parameter info and calling convention on first use

Reply:
I'm push another commit soon that updates this area of code.